File: VB/DataTable/ViewModels/MainViewModel.vb
Imports System
Imports System.Data
Imports DevExpress.Mvvm.CodeGenerators

Namespace DataTable.ViewModels

    <GenerateViewModel>
    Public Partial Class MainViewModel

        <GenerateProperty>
        Private _items As Data.DataTable

        Public Sub New()
            _items = New Data.DataTable()
            _items.Columns.Add(New DataColumn("Id", GetType(Integer)))
            _items.Columns.Add(New DataColumn("Name", GetType(String)))
            _items.Columns.Add(New DataColumn("CreatedAt", GetType(Date)))
            _items.Columns.Add(New DataColumn("Value", GetType(Integer), "Id * 10"))
            For i = 0 To 10 - 1
                _items.Rows.Add(i, $"Item {i}", Date.Today.AddDays(i))
            Next
        End Sub

        <GenerateCommand>
        Public Sub AddColumn()
            _items.Columns.Add(New DataColumn($"Value {_items.Columns.Count}", GetType(Integer), $"Id * {_items.Columns.Count * 10}"))
        End Sub
    End Class
End Namespace

File: CS/DataTable/ViewModels/MainViewModel.cs
using DevExpress.Mvvm;
using System;
using System.Data;
using System.Windows.Input;

namespace DataTable.ViewModels {
    public class MainViewModel : ViewModelBase {
        public System.Data.DataTable Items { get => GetProperty(() => Items); set => SetProperty(() => Items, value); }

        public MainViewModel() {
            Items = new System.Data.DataTable();

            Items.Columns.Add(new DataColumn("Id", typeof(int)));
            Items.Columns.Add(new DataColumn("Name", typeof(string)));
            Items.Columns.Add(new DataColumn("CreatedAt", typeof(DateTime)));
            Items.Columns.Add(new DataColumn("Value", typeof(int), "Id * 10"));

            for (var i = 0; i < 10; i++)
                Items.Rows.Add(i, $"Item {i}", DateTime.Today.AddDays(i));

            AddColumnCommand = new DelegateCommand(AddColumn);
        }

        public void AddColumn() {
            Items.Columns.Add(new DataColumn($"Value {Items.Columns.Count}", typeof(int),
                $"Id * {Items.Columns.Count * 10}"));
        }

        public ICommand AddColumnCommand { get; }
    }
}